pennylane/debugging/operation_checker.py

```diff
@@ -0,0 +1,833 @@
+
+    Returns:
+        bool: Whether the input matrix is a diagonal matrix
+    """
+    # Extract the diagonal, subtract it from the input, and check whether the result is 0.
+    off_diagonal = matrix - np.diag(np.diag(matrix))
+    return np.allclose(off_diagonal, np.zeros_like(matrix), atol=atol, rtol=0.0)
+
+
+def complex_jacobian(fn):
+    r_jac = qml.jacobian(lambda *args, **kwargs: qml.math.real(fn(*args, **kwargs)))
+    i_jac = qml.jacobian(lambda *args, **kwargs: qml.math.imag(fn(*args, **kwargs)))
+    return lambda *args, **kwargs: r_jac(*args, **kwargs) + 1j * i_jac(*args, **kwargs)
+
+
+def recipe_yields_commutator(recipe, op, par, wires):
+    """For now only implemented for single-parameter gates."""
+    coeffs, multipliers, shifts = recipe
+    shifted_parameters = [m * par[0] + s for m, s in zip(multipliers, shifts)]
+    shifted_matrices = [qml.matrix(op(p, wires=wires)) for p in shifted_parameters]
+
+    def apply_recipe(H):
+        out = qml.math.zeros_like(H)
+        for c, mat in zip(coeffs, shifted_matrices):
+            out = out + c * mat.T.conj() @ H @ mat
+        return out
+
+    d_apply_recipe = complex_jacobian(apply_recipe)
+
+    mat_fn = lambda p: qml.matrix(op(p, wires=wires))
+    d_mat_fn = complex_jacobian(mat_fn)
+    orig_mat = mat_fn(par[0])
+    d_orig_mat = d_mat_fn(par[0])
+    commutator_fn = (
+        lambda H: d_orig_mat.conj().T @ H @ orig_mat + orig_mat.conj().T @ H @ d_orig_mat
+    )
+    d_commutator_fn = complex_jacobian(commutator_fn)
+
+    dim = 2 ** len(wires)
+    H = (
+        np.random.random((dim, dim), requires_grad=True) * 2
+        + np.random.random((dim, dim), requires_grad=True) * 2j
+    )
+    H += H.T.conj()
+    return np.allclose(d_apply_recipe(H), d_commutator_fn(H))
+
+
+def wrap_op_method(op, method, expected_exc):
+    r"""Wrap a method of an operation with a try-except clause, allowing for an expected
+    exception and catching (and returning) other exceptions.
+
+    Args:
+        op (type or .operation.Operation): Operation that has the method to be wrapped
+        method (str): Name of the method to be wrapped
+        expected_exc (type): Exception type to ignore
+
+    Returns:
+        callable: The wrapped method of the operation.
+
+    The object returned by the returned callable differs, depending on the scenario:
+
+      - If the method succeeds, the return value of the method is returned,
+      - If the ``expected_exc`` is raised, ``None`` is returned,
+      - If another exception is raised, it is returned (but not raised).
+    """
+    _method = getattr(op, method)
+
+    def wrapped_method(*args, **kwargs):
+        r"""Wrapped operation method that tolerates an expected exception
+        and catches (and returns) all other exceptions."""
+        try:
+            return _method(*args, **kwargs)
+        except expected_exc:
+            return None
+        except Exception as e:
+            return e
+
+    return wrapped_method
```

Comment:
I might be wrong, but this feels like a use-case of https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager? Although I'm not 100% sure

Reply:
Oh wait! maybe this is what you are looking for: https://docs.python.org/3/library/contextlib.html#contextlib.suppress

Reply:
Nice, I wasn't aware of this context manager!
Maybe I'm overlooking an easy way of applying it here, but it seems that because we want to have three different return values (depending on whether no, an expected, or an unexpected error was raised) it's a bit tricky:
The `suppress` statement would only take care of the expected exception scenario, and the reason for returning `None` in that case would be quite subtle: `return _method(*args, **kwargs)` would fail and the default return value, `None`, would be returned.